src/doorman/test/TelnetProtocolTest.cpp

```diff
@@ -0,0 +1,270 @@
+#include "TelnetProtocol.hpp"
+
+#include <fmt/format.h>
+#include <fmt/ostream.h>
+
+#include <arpa/telnet.h>
+#include <catch2/catch.hpp>
+#include <catch2/trompeloeil.hpp>
+
+namespace {
+class MockHandler : public TelnetProtocol::Handler {
+public:
+    MAKE_MOCK1(send_bytes, void(gsl::span<const byte>), override);
+    MAKE_MOCK1(on_line, void(std::string_view), override);
+    MAKE_MOCK2(on_terminal_size, void(unsigned int, unsigned int), override);
+    MAKE_MOCK2(on_terminal_type, void(std::string_view type, bool ansi_supported), override);
+};
+
+using bytes = std::vector<byte>;
+auto wrap(std::string_view sv) {
+    bytes b;
+    for (auto c : sv)
+        b.emplace_back(static_cast<byte>(c));
+    return b;
+}
+auto match_bytes(const bytes &bytes) {
+    return trompeloeil::make_matcher<gsl::span<const byte>>(
+        [bytes](gsl::span<const byte> span) {
+            return span.size() == bytes.size() && memcmp(span.data(), bytes.data(), span.size()) == 0;
+        },
+        [bytes](std::ostream &os) { fmt::print(os, " is bytes{{{:02x}}}", fmt::join(bytes, ", ")); });
+}
+}
+
+TEST_CASE("Telnet protocol", "[TelnetProtocol]") {
+    MockHandler mock;
+    TelnetProtocol tp(mock);
+    SECTION("Should start out not supporting ANSI") { CHECK(!tp.supports_ansi()); }
+    SECTION("Should start out empty") { CHECK(tp.buffered_data_size() == 0); }
+    SECTION("Should send appropriate telnet requests and options on request") {
+        REQUIRE_CALL(mock, send_bytes(match_bytes(bytes{IAC, WONT, TELOPT_ECHO})));
```

Comment:
I see it writes the test failures to:  cmake-build-debug/Testing/Temporary/LastTest.log

Reply:
I see that Trompeloeil also has a TIMES and IN_SEQUENCE macro.  The Java equivalents would be:
```
verify(mockObject, times(3)).someMethod()
```
and
InOrder:
https://www.javadoc.io/doc/org.mockito/mockito-core/2.6.9/org/mockito/InOrder.html

Reply:
And never() is it's FORBID_CALL by the looks of things.

Reply:
Right: tromp forbids all by default, so I don't explicitly call that out here.

Reply:
> I see it writes the test failures to: cmake-build-debug/Testing/Temporary/LastTest.log

If you use `make test`, yeah (`cmake`'s idea of testing is overcomplex IMO).

IDE support usually is better, and if you run the tests directly you get prettier output (esp if they fail).

cmake's testing is test-system agnostic and you can even define tests that are bash scripts etc. It just knows it has to run all the things marked as test, and so to keep things simple it redirects output from each test to a file.